src/Week2/LogAnalysis.hs

```diff
@@ -0,0 +1,161 @@
+   parseMessage "I 29 la la la" == LogMessage Info 29 "la la la"
+   parseMessage "This is not in the right format" == Unknown "This is not in the right format"
+
+ Once we can parse one log message, we can parse a whole log file.
+ Define a function
+   parse :: String -> [LogMessage]
+ which parses an entire log file at once and returns its contents as a
+ list of LogMessages.
+
+ To test your function, use the testParse function provided in the
+ Log module, giving it as arguments your parse function, the number
+ of lines to parse, and the log file to parse from (which should also be
+ in the same folder as your assignment). For example, after loading
+ your assignment into GHCi, type something like this at the prompt:
+   testParse parse 10 "error.log"
+
+ Don’t reinvent the wheel! (That’s so last week.) Use Prelude functions to make
+ your solution as concise, high-level, and functional as
+ possible. For example, to convert a String like "562" into an Int, you
+ can use the read function. Other functions which may (or may not)
+ be useful to you include lines, words, unwords, take, drop, and (.).
+-}
+parseMessage :: String -> LogMessage
+parseMessage msg = case words msg of
+  ("I" : timestamp : info) -> case (readMaybe timestamp :: Maybe Int) > Just 0 of
+    True -> LogMessage Info (read timestamp) (unwords info)
+    _ -> Unknown msg
+  ("W" : timestamp : info) -> case (readMaybe timestamp :: Maybe Int) > Just 0 of
+    True -> LogMessage Warning (read timestamp) (unwords info)
+    _ -> Unknown msg
+  ("E" : level : timestamp : info) -> case (readMaybe level :: Maybe Int) > Just 0
+                                      && (readMaybe timestamp :: Maybe Int) > Just 0 of
+    True -> LogMessage (Error $ read level) (read timestamp) (unwords info)
+    _ -> Unknown msg
+  info -> Unknown (unwords info)
+
+
+-- Parse an entire log file at once and return its contents as a list of LogMessages
+parse :: String -> [LogMessage]
+parse = map (parseMessage) . lines
+
+
+-------- Putting the logs in order ------------
+
+{-
+ Exercise 2: Define a function
+    insert :: LogMessage -> MessageTree -> MessageTree
+
+ which inserts a new LogMessage into an existing MessageTree, producing a new MessageTree.
+ Insert may assume that it is given a sorted MessageTree, and must produce a new sorted
+ MessageTree containing the new LogMessage in addition to the contents of the
+ original MessageTree.
+
+ However, note that if insert is given a LogMessage which is Unknown, it should
+ return the MessageTree unchanged.
+-}
+insert :: LogMessage -> MessageTree -> MessageTree
+insert x Leaf = Node Leaf x Leaf
+insert (Unknown _) tree = tree
+insert _ tree@(Node _ (Unknown _) _) = tree
+insert msg1@(LogMessage _ t1 _ ) tree@(Node left msg2@(LogMessage _ t2 _) right)
+  | t2 > t1   = Node (insert msg1 left) msg2 right
+  | t2 < t1   = Node left msg2 (insert msg1 right)
+  | otherwise = tree
```

Comment:
Why would you not insert the `LogMessage` if `t1 = t2`? Is it not possible to have log messages with the same timestamp that differ in content?

Reply:
Oops. I missed the adding context of the exercise 2. Sorry about that. I've added missing parts.

It's given that the all the nodes in left subtree should be smaller and right subtree should have all nodes greater that the root.
Since, this is a BST there should not be a case of `t1==t2` as duplicate node values are not allowed.